feat(provider): detect venv python via "pynvim-python" tool

drmikehenry opened the pull request:
Problem: Detection of the pynvim module is currently done by finding the first Python interpreter in the `PATH` and checking if it can import pynvim.  This has several effects:

- Activation of an unrelated Python virtual environment will break automatic detection, unless pynvim is also installed in that environment.

- Installing pynvim to the expected location is difficult. User installation into the system-wide or user-wide Python site area is now deprecated.  On Ubuntu 24.04 with Python 3.12, for example, the command `pip install --user pynvim` now fails with the error message `error: externally-managed-environment`.

- Users may create a dedicated virtual environment in which to install pynvim, but Neovim won't detect it; instead, they must either activate it before launching Neovim (which interferes with the user of other virtual environments) or else hard-code the variable `g:python3_host_prog` in their `init.vim` to the path of the correct Python interpreter.  Neither option is desirable.

Solution: Expose pynvim's Python interpreter on the `PATH` under the name `pynvim-python`.  In the typical flow:

- User installs either uv or pipx.

- User installs pynvim via:

      uv tool install --upgrade pynvim
      # Or:
      pipx install --upgrade pynvim

With corresponding changes in pynvim (https://github.com/neovim/pynvim/pull/594), the above is all that's needed for Neovim to detect the installed location of pynvim, even if an unrelated Python virtual environments is activated.  It uses standard Python tooling to automate the necessary creation of a Python virtual environment for pyenv and the publication of `pynvim-python` to a directory on `PATH`.

See https://github.com/neovim/pynvim/issues/593 for additional discussion of this idea.

I've aimed to update all relevant portions of the documentation. I'm happy to make adjustments, or you may feel free to make changes as needed.

drmikehenry committed:
feat(provider): improve Neovim detection of pynvim

Problem: Detection of the pynvim module is currently done by finding the
first Python interpreter in the `PATH` and checking if it can import
pynvim.  This has several effects:

- Activation of an unrelated Python virtual environment will break
  automatic detection, unless pynvim is also installed in that
  environment.

- Installing pynvim to the expected location is difficult. User
  installation into the system-wide or user-wide Python site area is now
  deprecated.  On Ubuntu 24.04 with Python 3.12, for example, the
  command `pip install --user pynvim` now fails with the error message
  `error: externally-managed-environment`.

- Users may create a dedicated virtual environment in which to install
  pynvim, but Neovim won't detect it; instead, they must either activate
  it before launching Neovim (which interferes with the user of other
  virtual environments) or else hard-code the variable
  `g:python3_host_prog` in their `init.vim` to the path of the correct
  Python interpreter.  Neither option is desirable.

Solution: Expose pynvim's Python interpreter on the `PATH` under the
name `pynvim-python`.  In the typical flow:

- User installs either uv or pipx.

- User installs pynvim via:

      uv tool install --upgrade pynvim
      # Or:
      pipx install --upgrade pynvim

With corresponding changes in pynvim, the above is all that's needed for
Neovim to detect the installed location of pynvim, even if an unrelated
Python virtual environments is activated.  It uses standard Python
tooling to automate the necessary creation of a Python virtual
environment for pyenv and the publication of `pynvim-python` to a
directory on `PATH`.

See https://github.com/neovim/pynvim/issues/593 for additional
discussion of this idea.

justinmk commented on runtime/doc/provider.txt:
```diff
@@ -65,20 +71,21 @@ To disable Python 3 support: >vim
+interpreter and checking if it could import pynvim, automatic detection would
+fail when another virtual environment is active.  Upgrading to the latest
+pynvim is the recommended solution to this; but if that's not an option, then
+you can set the variable |python3_host_prog| in `init.vim` to point to the full
```
docs lint is failing bc of this
```suggestion
you can set the variable |g:python3_host_prog| in `init.vim` to point to the full
```

drmikehenry replied:
Added `g:` prefix.

justinmk commented on runtime/lua/vim/provider/python.lua:
```diff
@@ -83,6 +83,10 @@ function M.detect_by_module(module)
   end
 
+  if vim.fn.executable('pynvim-python') == 1 then
+    return 'pynvim-python'
```
`pynvim-python` isn't mentioned in provider.txt, I would expect some mention of it

justinmk reviewed (approved):
LGTM, after fixing the CI issues. This should be mentioned in news.txt

Reviewer note: this was thoroughly discussed in https://github.com/neovim/pynvim/issues/593 . Read that carefully before commenting.

drmikehenry commented:
Thanks for the review.  I've made the suggested changes (add `g:` to `python3_host_prog` and mention `pynvim-python` in `provider.txt`).  I've also updated the commit message to avoid starting with a capital letter. I've rebased onto `master` and squashed these changes into the branch.

I don't know how to add the feature into `runtime/doc/news.txt`, nor when in the process flow that should be added, so I didn't modify that file.

justinmk commented:
> I don't know how to tell what's wrong with the spelling in the below error message, nor how to reformat the text or allow the spelling:

change it from "Neovim" to "Nvim". this check is just for consistency in our docs and messages.

> how to add the feature into `runtime/doc/news.txt`

just add an item here that explains (concisely) what user-facing improvement we gained: https://github.com/neovim/neovim/blob/bd45e2be634b49f17b86a42359c0218081759f48/runtime/doc/news.txt#L275

justinmk commented on runtime/doc/provider.txt:
```diff
@@ -36,19 +36,25 @@ itself).
-   python3 -m pip install --user --upgrade pynvim
-
-The pip `--upgrade` flag ensures that you get the latest version even if
+2. Make sure either uv (https://docs.astral.sh/uv/) or pipx
+   (https://pipx.pypa.io/stable/) is installed.
```
```suggestion
2. Install either uv (https://docs.astral.sh/uv/) or pipx
   (https://pipx.pypa.io/stable/).
```

justinmk commented on runtime/doc/provider.txt:
```diff
@@ -36,19 +36,25 @@ itself).
 See also |python-virtualenv|.
 
 Note: The old "neovim" module was renamed to "pynvim".
 https://github.com/neovim/neovim/wiki/Following-HEAD#20181118
 If you run into problems, uninstall _both_ then install "pynvim" again: >bash
+  # Note: this step depends on how you originally installed these:
   python -m pip uninstall neovim pynvim
-  python -m pip install --user --upgrade pynvim
+  # Install pynvim as above, e.g.:
+  uv tool install --upgrade pynvim
```
can probably drop this whole passage now, it has been ages :)

drmikehenry commented on runtime/doc/provider.txt:
```diff
@@ -65,20 +71,22 @@ To disable Python 3 support: >vim
+  let g:python3_host_prog = '/path/to/pynvim-venv/bin/python'
+
+  " For Windows:
+  let g:python3_host_prog = 'c:\path\to\pynvim-venv\bin\python.exe'
```
Adjusted as suggested.

justinmk commented:
@drmikehenry Thanks again for your persistence here. This technique is lightweight and high-leverage and may even be a useful reference for other providers.

drmikehenry commented:
@justinmk I've spent a few hours looking at the Python provider health checking code, and it's fairly involved.  While I suspect you're right that simplification is possible, I don't know which of the checks are considered valuable and which are obsolete, and unfortunately reworking this logic is more than I can take on at this time.

Regarding the lines you linked, I agree the text is obsolete:
https://github.com/neovim/neovim/blob/bd45e2be634b49f17b86a42359c0218081759f48/runtime/lua/vim/provider/health.lua#L734-L739

But I think it's pretty unlikely for the advice to be displayed, as it looks to me like it happens when `import neovim` fails but `import pynvim` succeeds.  That sounds like logic from the distant past when the `neovim` module was renamed to `pynvim`, so probably nobody will encounter this error message.

The advice here seems more likely to be displayed, though I've not been able to trigger it to happen:
https://github.com/neovim/neovim/blob/bd45e2be634b49f17b86a42359c0218081759f48/runtime/lua/vim/provider/health.lua#L761-L766

Changing the advice to point to `:help provider` or to offer `uv tool install --upgrade pynvim` may be worth doing, if you understand the scenario where the advice would be displayed.

justinmk commented:
> Changing the advice to point to `:help provider` or to offer `uv tool install --upgrade pynvim` may be worth doing,

Thanks, yeah that is probably good enough for now. 

>  if you understand the scenario where the advice would be displayed.

If pynvim isn't found it will/should be triggered.

drmikehenry commented:
When I have no `pynvim` installed and launch `nvim --clean` using either Neovim 0.11.3 or Neovim `master`, then run:

```vim
:checkhealth vim.provider
```

I get this Python-related output that does not contain the message `pynvim is not installed`:
```k
Python 3 provider (optional) ~
- ⚠️ WARNING No Python executable found that can `import neovim`. Using the first available executable for diagnostics.
- ⚠️ WARNING Could not load Python :
  /usr/bin/python3 does not have the "neovim" module.
  /home/mike/.local/bin/python3.13 does not have the "neovim" module.
  /usr/bin/python3.12 does not have the "neovim" module.
  python3.11 not found in search path or not executable.
  python3.10 not found in search path or not executable.
  python3.9 not found in search path or not executable.
  /usr/local/bin/python does not have the "neovim" module.
  - ADVICE:
    - See :help |provider-python| for more information.
    - You can disable this provider (and warning) by adding `let g:loaded_python3_provider = 0` to your init.vim
- Executable: Not found

Python virtualenv ~
- ✅ OK no $VIRTUAL_ENV
```

There may be a more complicated path through the checks that results in displaying this advice.  It certainly has the feel of a plain "pynvim isn't installed" message, but just lacking `pynvim` isn't enough to trigger the message.

justinmk commented:
ok. the healthcheck for python is complicated and your change gives us some hope of simplifying it, but meantime just updating any relevant message strings is good enough.
